go/vt/vtgate/evalengine/fn_string.go

```diff
@@ -932,6 +937,105 @@ func (call builtinTrim) compile(c *compiler) (ctype, error) {
 	return ctype{Type: sqltypes.VarChar, Flag: flagNullable, Col: col}, nil
 }
 
+func (call *builtinSubstring) eval(env *ExpressionEnv) (eval, error) {
+	str, err := call.Arguments[0].eval(env)
+	if err != nil || str == nil {
+		return nil, err
+	}
+
+	text, ok := str.(*evalBytes)
+	tt := str.SQLType()
+	if !ok {
+		text, err = evalToVarchar(str, call.collate, true)
+		if err != nil {
+			return nil, err
+		}
+		tt = sqltypes.VarChar
+	}
+
+	p, err := call.Arguments[1].eval(env)
+	if err != nil || p == nil {
```

Comment:
I'm not aiming to try to fix this here all at once, that's maybe something for another PR but I vastly prefer to keep consistent styling here for now.

Reply:
> Ah, seeing this same pattern throughout. It may potentially be a good cleanup to have the eval() path return errUndefined or similar as an error that can be handled specially rather than making the caller check both returns at each call site.

Also the logic here needs to be call site dependent to match how MySQL behaves. So we need to check both sometimes, sometimes one etc. Also I'm not really sure what here `errUndefined` would mean, since `NULL` in SQL (and `nil` here) is a valid value in common cases anyway. So I suspect it might even make it harder to follow tbh to change this.